source/adapters/cuda/command_buffer.cpp

```diff
@@ -96,6 +122,28 @@ ur_exp_command_buffer_command_handle_t_::
     std::memset(GlobalWorkOffset + WorkDim, 0, ZeroSize);
     std::memset(GlobalWorkSize + WorkDim, 0, ZeroSize);
   }
+};
+
+ur_exp_command_buffer_command_handle_t_::
+    ur_exp_command_buffer_command_handle_t_(
+        ur_exp_command_buffer_handle_t CommandBuffer, CUgraphNode Node,
+        CUgraphNode SignalNode, std::vector<CUgraphNode> WaitNodes)
+    : CommandBuffer(CommandBuffer), Node(Node), SignalNode(SignalNode),
+      WaitNodes(WaitNodes), RefCountInternal(1), RefCountExternal(1) {
+  CommandBuffer->incrementInternalReferenceCount();
+}
+
+ur_exp_command_buffer_command_handle_t_::
+    ~ur_exp_command_buffer_command_handle_t_() {
+  // We create the ur_event_t returned to the user for a signal node using
+  // `makeWithNative` which sets `HasOwnership` to false. Therefore destruction
+  // of the `ur_event_t` object doesn't free the underlying CuEvent_t object and
+  // we need to do it manually ourselves.
+  if (SignalNode) {
+    CUevent SignalEvent;
+    cuGraphEventRecordNodeGetEvent(SignalNode, &SignalEvent);
```

Comment:
Should we be checking for an error here?

Reply:
These CUDA calls were inside a destructor, so throwing exception with `UR_CHECK_ERROR` would be bad form. So i've moved the CUDA calls just outside of the destructor so that we can do this checking with `UR_CHECK_ERROR`

Reply:
Nice thanks